File: 10178-383-547-number-of-provinces/10178-383-547-number-of-provinces.js
/**
 * @param {number[][]} isConnected
 * @return {number}
 */
var findCircleNum = function(isConnected) {
    let count = 0,visited = Array(isConnected.length).fill(false)

    const dfs = key =>{
        if(!visited[key]){
            visited[key] = true
            for(let i = 0;i<isConnected[key].length;i++){
                if(isConnected[key][i]===1){
                    dfs(i)
                }
            }
        }
    }
    for(let i = 0;i<visited.length;i++){
        if(!visited[i]){
            count++
            dfs(i)
        }
    }
    return count
};

File: 10178-383-547-number-of-provinces/10178-383-547-number-of-provinces.java
class Solution {
    
    public int findCircleNum(int[][] isConnected) {
        boolean visited[] = new boolean[isConnected.length];
        int count = 0;
        for(int i=0;i<visited.length;i++){
            if(!visited[i]){
                count++;
                dfs(isConnected,i,visited);
            }
        }
        return count;
    }

    public void dfs(int[][] isConnected,int i,boolean visited[]){
        if(!visited[i]){
            visited[i] = true;
            for(int j=0;j<isConnected[i].length;j++){
                if(isConnected[i][j]==1 && !visited[j]){
                    dfs(isConnected,j,visited);
                }
            }
        }
    }
}